### DISCUSSION TITLE
I cannot understand the function about “OpenAPI Callbacks”

### DISCUSSION THREAD

**Discussion by mht2953658596** [fastapi/fastapi#10384] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Union

from fastapi import APIRouter, FastAPI
from pydantic import BaseModel, HttpUrl

app = FastAPI()


class Invoice(BaseModel):
    id: str
    title: Union[str, None] = None
    customer: str
    total: float


class InvoiceEvent(BaseModel):
    description: str
    paid: bool


class InvoiceEventReceived(BaseModel):
    ok: bool


invoices_callback_router = APIRouter()


@invoices_callback_router.post(
    "{$callback_url}/invoices/{$request.body.id}", response_model=InvoiceEventReceived
)
def invoice_notification(body: InvoiceEvent):
    pass


@app.post("/invoices/", callbacks=invoices_callback_router.routes)
def create_invoice(invoice: Invoice, callback_url: Union[HttpUrl, None] = None):
    """
    Create an invoice.

    This will (let's imagine) let the API user (some external developer) create an
    invoice.

    And this path operation will:

    * Send the invoice to the client.
    * Collect the money from the client.
    * Send a notification back to the API user (the external developer), as a callback.
        * At this point is that the API will somehow send a POST request to the
            external API with the notification of the invoice event
            (e.g. "payment successful").
    """
    # Send the invoice, collect the money, send the notification (the callback)
    return {"msg": "Invoice received"}
```

### Description

While reading this document, I have always been unable to understand the function of OpenAPI Callbacks.

My understanding is that when a client accesses the API endpoint of my FastAPI application, my API can make requests to external APIs, such as another programmer's application interface created using Spring Boot, which is located at https://www.external.org/events.

In other words, when a client accesses the API I've written, let's call it "A API," this "A API" will make requests to another API, which we'll call "B API."

My question is:
When I look at the examples in the documentation,
```python

@invoices_callback_router.post(
    "{$callback_url}/invoices/{$request.body.id}", response_model=InvoiceEventReceived
)
def invoice_notification(body: InvoiceEvent):
    pass
```

My understanding is that when I access the /invoices/ endpoint and provide the callback_url variable with https://www.external.org/events, along with a JSON body of:
```
{
"id": "2expen51ve",
"customer": "Mr. Richie Rich",
"total": "9999"
}
```
Will the /invoices/ endpoint call https://www.external.org/events/invoices/2expen51ve? Is that correct?

However, in the line of code `def invoice_notification(body: InvoiceEvent):`, don't we need to pass a JSON similar to this:
```
{
"description": "Payment celebration",
"paid": true
}  
```
？
Is this JSON being passed from the create_invoice function to the invoice_notification function?

Or, where does this JSON originate from?Where is this JSON input from?

```
{
"description": "Payment celebration",
"paid": true
}  
```

This is the first question.

Another question is,

Suppose I have written two pieces of code,

```python
@invoices_callback_router.post(
	"{$callback_url}/invoices/{$request.body.id}", response_model=InvoiceEventReceived
)
def invoice_notification_1(body: InvoiceEvent):
	pass


@invoices_callback_router.post(
	"{$callback_url}/invoices/{$request.body.title}", response_model=InvoiceEventReceived
)
def invoice_notification_2(body: InvoiceEvent):
	pass

```

Does this mean that when a client accesses my FastAPI's A API endpoint, this A API will simultaneously make requests to both the B API and C API endpoints?

### Operating System

Windows

### FastAPI Version

0.83.0

### Pydantic Version

1.10.2

### Python Version

Python 3.8.5

**Comment by jgould22** [comment, marked as answer]
From https://fastapi.tiangolo.com/advanced/openapi-callbacks/

This page is all about **documenting** to your clients what you intend to do, it only documents, you have to create the implementation yourself.

> Will the /invoices/ endpoint call https://www.external.org/events/invoices/2expen51ve? Is that correct?

Not automatically, you have to call it yourself, it will do whatever you code it to do.

The rest of the page defining the routers etc is there so fastapi can generate this OpenAPI schema for the swagger doc.

**Reply by mht2953658596** [reply]
I did read the official documentation you pointed out, but I still can't understand this callback.
According to your statement, 

> 'Not automatically, you have to call it yourself, it will do whatever you code it to do,'

 why don't I write the code like this instead?
```python
from typing import Union
import httpx
from fastapi import APIRouter, FastAPI
from pydantic import BaseModel, HttpUrl

app = FastAPI()

class Invoice(BaseModel):
    id: str
    title: Union[str, None] = None
    customer: str
    total: float

class InvoiceEvent(BaseModel):
    description: str
    paid: bool

class InvoiceEventReceived(BaseModel):
    ok: bool

invoices_callback_router = APIRouter()

@invoices_callback_router.post(
    "/invoices/{invoice_id}/callback", response_model=InvoiceEventReceived
)
async def invoice_notification(invoice_id: str, body: InvoiceEvent):
    return {"ok": True}

@app.post("/invoices/", response_model=Invoice)
async def create_invoice(invoice: Invoice, callback_url: Union[HttpUrl, None] = None):


    if callback_url:
        event_data = {
            "description": "Payment successful",
            "paid": True
        }
        try:
            async with httpx.AsyncClient() as client:
                response = await client.post(callback_url, json=event_data)
            response.raise_for_status()  
        except httpx.HTTPError as e:
            return {"error": str(e)}

    return invoice

```

Can't I just call the '/invoices/{invoice_id}/callback' API directly within the '/invoices/' API endpoint?
What does the setting `callbacks=invoices_callback_router.routes` in the official documentation mean in the following code?

```python
@app.post("/invoices/", callbacks=invoices_callback_router.routes)
def create_invoice(invoice: Invoice, callback_url: Union[HttpUrl, None] = None):
    return {"msg": "Invoice received"}
```


So, I still can't understand. If you don't mind, could you please provide a demo code to explain it?

**Reply by jgould22** [reply]
>Can't I just call the '/invoices/{invoice_id}/callback' API directly within the '/invoices/' API endpoint?

Yes, that is what you are supposed to do fastapi does not do this for you.

>What does the setting callbacks=invoices_callback_router.routes in the official documentation mean in the following code?

It tells FastAPI to look at that router and put information it finds there into the callback tab for that request as shown in the docs picture here https://fastapi.tiangolo.com/advanced/openapi-callbacks/#check-the-docs.  You can see the callbacks tab beside the parameters tab in the swagger doc. 

That is all it does, this is about documentation.

If that does not make it clear perhaps you can tell me how you think this feature should work and I can figure out what information you might be missing.

**Reply by mht2953658596** [reply]
So, this section on callbacks is just telling us how to provide more informative details in Swagger UI? Is that all? I thought it could automatically call another B API interface when using an A API interface. It seems that this "callback" feature is only for providing additional hints.

**Reply by jgould22** [reply]
>So, this section on callbacks is just telling us how to provide more informative details in Swagger UI?

Correct, it is a way to that allows FastAPI to generate part of the open api using python.

**Reply by mht2953658596** [reply]
```python
@invoices_callback_router.post(
	"{$callback_url}/invoices/{$request.body.id}", response_model=InvoiceEventReceived
)
def invoice_notification_1(body: InvoiceEvent):
	pass


@invoices_callback_router.post(
	"{$callback_url}/invoices/{$request.body.title}", response_model=InvoiceEventReceived
)
def invoice_notification_2(body: InvoiceEvent):
	pass

```

Also, there is another question I hope you can answer: if I have defined two external API interfaces

Will there be a conflict when setting @app.post("/invoices/", callbacks=invoices_callback_router.routes)?
